File: scripts/generate_pre_match_data.js
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Set random seed for reproducibility
function seededRandom(seed) {
    const x = Math.sin(seed++) * 10000;
    return x - Math.floor(x);
}

let random = seededRandom.bind(null, 42);

function getTeamNames() {
    const databasePath = "my-world-cup/database/2_ability_models";
    if (!fs.existsSync(databasePath)) {
        throw new Error(`Database directory not found at ${databasePath}`);
    }

    // Get all directories that look like team names
    const teamNames = [];
    const items = fs.readdirSync(databasePath);

    for (const item of items) {
        const itemPath = path.join(databasePath, item);
        if (fs.statSync(itemPath).isDirectory()) {
            teamNames.push(item);
        }
    }

    // Get 3-letter team codes from team_names
    const teamCodes = [];
    for (const name of teamNames) {
        if (name.includes(' ')) {
            const parts = name.split(' ');
            const code = parts[parts.length - 1];
            teamCodes.push(code);
        } else {
            teamCodes.push(name.substring(0, 3).toUpperCase());
        }
    }

    return teamCodes;
}

function getTeamOverallScore(teamName) {
    const radarFile = `my-world-cup/database/2_ability_models/${teamName}/radar_data.json`;

    if (fs.existsSync(radarFile)) {
        const data = fs.readFileSync(radarFile, 'utf8');
        const radarData = JSON.parse(data);

        // Calculate overall as average of all attributes
        const attributes = ['attack', 'defense', 'control', 'status', 'experience', 'place_kick', 'superstar', 'penalty'];
        const values = attributes.map(attr => radarData[attr] || 75);
        return Math.round(values.reduce((a, b) => a + b, 0) / values.length);
    }

    return 75; // Default if not found
}

function generateStrategy(teamId, overallScore) {
    // Determine possible styles based on overall score
    let possibleStyles;
    if (overallScore > 85) {
        possibleStyles = ['possession', 'pressing'];
    } else if (overallScore >= 75) {
        possibleStyles = ['balanced', 'possession', 'pressing'];
    } else if (overallScore >= 65) {
        possibleStyles = ['balanced', 'counter_attack', 'defensive'];
    } else {
        possibleStyles = ['defensive', 'counter_attack'];
    }

    const style = possibleStyles[Math.floor(random() * possibleStyles.length)];

    // Map styles to attributes
    const styleAttributes = {
        'possession': {
            pressing_intensity: ['medium', 'high'][Math.floor(random() * 2)],
            defensive_line: ['mid', 'high'][Math.floor(random() * 2)],
            transition_speed: ['medium', 'slow'][Math.floor(random() * 2)]
        },
        'counter_attack': {
            pressing_intensity: ['low', 'medium'][Math.floor(random() * 2)],
            defensive_line: ['low', 'mid'][Math.floor(random() * 2)],
            transition_speed: ['fast', 'medium'][Math.floor(random() * 2)]
        },
        'pressing': {
            pressing_intensity: 'high',
            defensive_line: ['high', 'mid'][Math.floor(random() * 2)],
            transition_speed: ['fast', 'medium'][Math.floor(random() * 2)]
        },
        'defensive': {
            pressing_intensity: 'low',
            defensive_line: 'low',
            transition_speed: ['slow', 'medium'][Math.floor(random() * 2)]
        },
        'balanced': {
            pressing_intensity: 'medium',
            defensive_line: 'mid',
            transition_speed: 'medium'
        }
    };

    // Formation weights by style
    const formationWeights = {
        'possession': {'4-3-3': 0.5, '4-2-3-1': 0.3, '3-4-3': 0.2},
        'pressing': {'4-3-3': 0.6, '4-4-2': 0.3, '3-5-2': 0.1},
        'defensive': {'5-3-2': 0.5, '4-4-2': 0.3, '4-5-1': 0.2},
        'counter_attack': {'4-2-3-1': 0.5, '4-3-3': 0.3, '3-4-3': 0.2},
        'balanced': {'4-3-3': 0.4, '4-2-3-1': 0.4, '4-4-2': 0.2}
    };

    const formations = Object.keys(formationWeights[style]);
    const weights = Object.values(formationWeights[style]);
    let formation = formations[0];
    let weightSum = 0;
    const r = random();

    for (let i = 0; i < weights.length; i++) {
        weightSum += weights[i];
        if (r <= weightSum) {
            formation = formations[i];
            break;
        }
    }

    return {
        team_id: teamId,
        formation: formation,
        style: style,
        pressing_intensity: styleAttributes[style].pressing_intensity,
        defensive_line: styleAttributes[style].defensive_line,
        transition_speed: styleAttributes[style].transition_speed
    };
}

function generateRecentForm(teamId, overallScore) {
    // Determine W/D/L distribution based on overall score
    let wins, draws, losses;

    if (overallScore > 85) {
        wins = Math.floor(random() * 3) + 7; // 7-9
        draws = Math.floor(random() * 3); // 0-2
    } else if (overallScore >= 75) {
        wins = Math.floor(random() * 3) + 5; // 5-7
        draws = Math.floor(random() * 3) + 1; // 1-3
    } else if (overallScore >= 65) {
        wins = Math.floor(random() * 3) + 3; // 3-5
        draws = Math.floor(random() * 3) + 2; // 2-4
    } else {
        wins = Math.floor(random() * 3) + 1; // 1-3
        draws = Math.floor(random() * 3) + 2; // 2-4
    }

    losses = Math.max(0, 10 - wins - draws);

    // Create result list
    const results = [];
    const opponents = [
        "巴西", "法国", "英格兰", "德国", "西班牙", "意大利", "葡萄牙", "荷兰",
        "克罗地亚", "阿根廷", "乌拉圭", "哥伦比亚", "墨西哥", "美国", "日本",
        "韩国", "澳大利亚", "智利", "秘鲁", "厄瓜多尔", "喀麦隆", "尼日利亚",
        "塞内加尔", "加纳", "摩洛哥", "阿尔及利亚", "突尼斯", "埃及", "沙特",
        "伊朗", "伊拉克", "阿联酋", "约旦", "黎巴嫩", "巴勒斯坦", "叙利亚",
        "乌兹别克斯坦", "哈萨克斯坦", "泰国", "越南", "印尼", "马来西亚", "新加坡"
    ];

    const resultTypes = [];
    for (let i = 0; i < wins; i++) resultTypes.push('W');
    for (let i = 0; i < draws; i++) resultTypes.push('D');
    for (let i = 0; i < losses; i++) resultTypes.push('L');

    // Make sure we have exactly 10 results
    while (resultTypes.length < 10) {
        resultTypes.push('L');
    }
    while (resultTypes.length > 10) {
        resultTypes.pop();
    }

    // Shuffle results
    for (let i = resultTypes.length - 1; i > 0; i--) {
        const j = Math.floor(random() * (i + 1));
        [resultTypes[i], resultTypes[j]] = [resultTypes[j], resultTypes[i]];
    }

    let goalsScored = 0;
    let goalsConceded = 0;

    // Use a copy of opponents to avoid picking the same opponent too often
    const availableOpponents = [...opponents];

    for (let i = 0; i < resultTypes.length; i++) {
        const result = resultTypes[i];
        // Pick opponent from remaining options, or from all if we've used them all
        const opponentIndex = Math.floor(random() * availableOpponents.length);
        let opponent = availableOpponents[opponentIndex];

        // Remove opponent if we have more games to play
        if (availableOpponents.length > 10) {
            availableOpponents.splice(opponentIndex, 1);
        } else {
            // If we've run out of unique opponents, shuffle and reset
            if (i > 0 && availableOpponents.length === 0) {
                availableOpponents.push(...opponents);
            }
        }

        // Score generation based on result
        let score, scored, conceded;
        if (result === 'W') {
            scored = Math.floor(random() * 3) + 2; // 2-4
            conceded = Math.floor(random() * 2); // 0-1
            score = `${scored}-${conceded}`;
        } else if (result === 'D') {
            scored = Math.floor(random() * 2) + 1; // 1-2
            conceded = scored;
            score = `${scored}-${conceded}`;
        } else { // Loss
            scored = Math.floor(random() * 3); // 0-2
            conceded = Math.floor(random() * 3) + 1; // 1-3
            score = `${scored}-${conceded}`;
        }

        goalsScored += scored;
        goalsConceded += conceded;

        results.push({
            opponent: opponent,
            result: result,
            score: score,
            type: random() > 0.5 ? "friendly" : "qualifier"
        });
    }

    return {
        team_id: teamId,
        last_10: results,
        summary: {
            wins: wins,
            draws: draws,
            losses: losses,
            goals_scored: goalsScored,
            goals_conceded: goalsConceded
        }
    };
}

function extendOddsFile() {
    const oddsFile = "my-world-cup/database/2_ability_models/odds.json";

    if (!fs.existsSync(oddsFile)) {
        console.log(`Warning: odds.json not found at ${oddsFile}`);
        return;
    }

    console.log(`Found odds.json at ${oddsFile}`);

    const data = fs.readFileSync(oddsFile, 'utf8');
    const oddsData = JSON.parse(data);

    // Extract odds values for normalization
    const oddsValues = oddsData.map(entry => entry.odds);
    const minOdds = Math.min(...oddsValues);
    const maxOdds = Math.max(...oddsValues);

    // Add new fields to each entry
    for (const entry of oddsData) {
        const strengthNorm = 1 - (entry.odds - minOdds) / (maxOdds - minOdds);

        const winOddsBase = Math.round((1.5 + (1 - strengthNorm) * 7.5) * 100) / 100;
        const drawOddsBase = Math.round((3.0 + Math.abs(strengthNorm - 0.5) * 0.8 + (random() - 0.5) * 0.2) * 100) / 100;
        const lossOddsBase = Math.round((1.2 + strengthNorm * 6.0) * 100) / 100;

        entry.win_odds_base = winOddsBase;
        entry.draw_odds_base = drawOddsBase;
        entry.loss_odds_base = lossOddsBase;
    }

    // Write back to file
    fs.writeFileSync(oddsFile, JSON.stringify(oddsData, null, 2), 'utf8');
    console.log(`Extended ${oddsData.length} entries in odds.json`);
}

function main() {
    console.log("Generating pre-match data...");

    try {
        // Get team names
        const teamCodes = getTeamNames();
        console.log(`Found ${teamCodes.length} teams: ${teamCodes.join(', ')}`);

        // Generate strategy.json and recent_form.json for each team
        for (const teamId of teamCodes) {
            // Find team name directory
            let teamNameDir = null;
            const items = fs.readdirSync("my-world-cup/database/2_ability_models");

            for (const item of items) {
                const itemPath = path.join("my-world-cup/database/2_ability_models", item);
                if (fs.statSync(itemPath).isDirectory()) {
                    // Check if this directory corresponds to our team
                    let code;
                    if (item.includes(' ')) {
                        const parts = item.split(' ');
                        code = parts[parts.length - 1];
                    } else {
                        code = item.substring(0, 3).toUpperCase();
                    }

                    if (code === teamId) {
                        teamNameDir = item;
                        break;
                    }
                }
            }

            if (!teamNameDir) {
                console.log(`Warning: Could not find directory for team ${teamId}`);
                continue;
            }

            // Get overall score
            const overallScore = getTeamOverallScore(teamNameDir);

            // Generate strategy
            const strategy = generateStrategy(teamId, overallScore);
            const strategyFile = `my-world-cup/database/2_ability_models/${teamNameDir}/strategy.json`;

            fs.writeFileSync(strategyFile, JSON.stringify(strategy, null, 2), 'utf8');

            // Generate recent form
            const recentForm = generateRecentForm(teamId, overallScore);
            const formFile = `my-world-cup/database/2_ability_models/${teamNameDir}/recent_form.json`;

            fs.writeFileSync(formFile, JSON.stringify(recentForm, null, 2), 'utf8');

            console.log(`Generated data for ${teamId} (${teamNameDir}): overall=${overallScore}`);
        }

        // Extend odds file
        extendOddsFile();

        console.log("Pre-match data generation complete!");
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

File: scripts/generate_pre_match_data.py
import json
import random
import os
import glob
from typing import Dict, List, Any

# Set random seed for reproducibility
random.seed(42)

def get_team_names():
    """Get list of all team names from database directory"""
    database_path = "my-world-cup/database/2_ability_models"
    if not os.path.exists(database_path):
        raise FileNotFoundError(f"Database directory not found at {database_path}")

    # Get all directories that look like team names
    team_names = []
    for item in os.listdir(database_path):
        item_path = os.path.join(database_path, item)
        if os.path.isdir(item_path):
            team_names.append(item)

    # Get 3-letter team codes from team_names (last part if space separated)
    team_codes = []
    for name in team_names:
        if ' ' in name:
            code = name.split()[-1]
        else:
            code = name[:3].upper()
        team_codes.append(code)

    return team_codes

def get_team_overall_score(team_name: str) -> int:
    """Get overall score for a team"""
    radar_file = f"my-world-cup/database/2_ability_models/{team_name}/radar_data.json"
    if os.path.exists(radar_file):
        with open(radar_file, 'r', encoding='utf-8') as f:
            radar_data = json.load(f)
            # Calculate overall as average of all attributes
            attributes = ['attack', 'defense', 'control', 'status', 'experience',
                         'place_kick', 'superstar', 'penalty']
            values = [radar_data.get(attr, 75) for attr in attributes]
            return int(round(sum(values) / len(values)))
    return 75  # Default if not found

def generate_strategy(team_id: str, overall_score: int) -> Dict[str, Any]:
    """Generate strategy data for a team based on overall score"""

    # Determine possible styles based on overall score
    if overall_score > 85:
        possible_styles = ['possession', 'pressing']
    elif overall_score >= 75:
        possible_styles = ['balanced', 'possession', 'pressing']
    elif overall_score >= 65:
        possible_styles = ['balanced', 'counter_attack', 'defensive']
    else:
        possible_styles = ['defensive', 'counter_attack']

    style = random.choice(possible_styles)

    # Map styles to attributes
    style_attributes = {
        'possession': {
            'pressing_intensity': random.choice(['medium', 'high']),
            'defensive_line': random.choice(['mid', 'high']),
            'transition_speed': random.choice(['medium', 'slow'])
        },
        'counter_attack': {
            'pressing_intensity': random.choice(['low', 'medium']),
            'defensive_line': random.choice(['low', 'mid']),
            'transition_speed': random.choice(['fast', 'medium'])
        },
        'pressing': {
            'pressing_intensity': 'high',
            'defensive_line': random.choice(['high', 'mid']),
            'transition_speed': random.choice(['fast', 'medium'])
        },
        'defensive': {
            'pressing_intensity': 'low',
            'defensive_line': 'low',
            'transition_speed': random.choice(['slow', 'medium'])
        },
        'balanced': {
            'pressing_intensity': 'medium',
            'defensive_line': 'mid',
            'transition_speed': 'medium'
        }
    }

    # Random formations weighted by style
    formation_weights = {
        'possession': {'4-3-3': 0.5, '4-2-3-1': 0.3, '3-4-3': 0.2},
        'pressing': {'4-3-3': 0.6, '4-4-2': 0.3, '3-5-2': 0.1},
        'defensive': {'5-3-2': 0.5, '4-4-2': 0.3, '4-5-1': 0.2},
        'counter_attack': {'4-2-3-1': 0.5, '4-3-3': 0.3, '3-4-3': 0.2},
        'balanced': {'4-3-3': 0.4, '4-2-3-1': 0.4, '4-4-2': 0.2}
    }

    formations = list(formation_weights[style].keys())
    weights = list(formation_weights[style].values())
    formation = random.choices(formations, weights=weights)[0]

    return {
        "team_id": team_id,
        "formation": formation,
        "style": style,
        "pressing_intensity": style_attributes[style]['pressing_intensity'],
        "defensive_line": style_attributes[style]['defensive_line'],
        "transition_speed": style_attributes[style]['transition_speed']
    }

def generate_recent_form(team_id: str, overall_score: int) -> Dict[str, Any]:
    """Generate recent form data for a team"""

    # Determine W/D/L distribution based on overall score
    if overall_score > 85:
        wins = random.randint(7, 9)
        draws = random.randint(0, 2)
        losses = 10 - wins - draws
    elif overall_score >= 75:
        wins = random.randint(5, 7)
        draws = random.randint(1, 3)
        losses = 10 - wins - draws
    elif overall_score >= 65:
        wins = random.randint(3, 5)
        draws = random.randint(2, 4)
        losses = 10 - wins - draws
    else:
        wins = random.randint(1, 3)
        draws = random.randint(2, 4)
        losses = 10 - wins - draws

    # Create result list
    results = []
    opponents = [
        "巴西", "法国", "英格兰", "德国", "西班牙", "意大利", "葡萄牙", "荷兰",
        "克罗地亚", "阿根廷", "乌拉圭", "哥伦比亚", "墨西哥", "美国", "日本",
        "韩国", "澳大利亚", "智利", "秘鲁", "厄瓜多尔", "喀麦隆", "尼日利亚",
        "塞内加尔", "加纳", "摩洛哥", "阿尔及利亚", "突尼斯", "埃及", "沙特",
        "伊朗", "伊拉克", "阿联酋", "约旦", "黎巴嫩", "巴勒斯坦", "叙利亚",
        "乌兹别克斯坦", "哈萨克斯坦", "泰国", "越南", "印尼", "马来西亚", "新加坡"
    ]

    result_types = ['W'] * wins + ['D'] * draws + ['L'] * losses
    random.shuffle(result_types)

    goals_scored = 0
    goals_conceded = 0

    for i, result in enumerate(result_types):
        opponent = random.choice(opponents)
        # Score generation based on result
        if result == 'W':
            score = f"{random.randint(2, 4)}-{random.randint(0, 1)}"
            scored, conceded = map(int, score.split('-'))
        elif result == 'D':
            score = f"{random.randint(1, 2)}-{random.randint(1, 2)}"
            scored, conceded = map(int, score.split('-'))
        else:  # Loss
            score = f"{random.randint(0, 2)}-{random.randint(1, 3)}"
            scored, conceded = map(int, score.split('-'))

        goals_scored += scored
        goals_conceded += conceded

        result_data = {
            "opponent": opponent,
            "result": result,
            "score": score,
            "type": random.choice(["friendly", "qualifier"])
        }
        results.append(result_data)

    return {
        "team_id": team_id,
        "last_10": results,
        "summary": {
            "wins": wins,
            "draws": draws,
            "losses": losses,
            "goals_scored": goals_scored,
            "goals_conceded": goals_conceded
        }
    }

def extend_odds_file():
    """Extend odds.json with new fields"""
    odds_file = "my-world-cup/database/2_ability_models/odds.json"

    if not os.path.exists(odds_file):
        print(f"Warning: odds.json not found at {odds_file}")
        return

    print(f"Found odds.json at {odds_file}")

    with open(odds_file, 'r', encoding='utf-8') as f:
        odds_data = json.load(f)

    # Extract odds values for normalization
    odds_values = [entry['odds'] for entry in odds_data]
    min_odds = min(odds_values)
    max_odds = max(odds_values)

    # Add new fields to each entry
    for entry in odds_data:
        strength_norm = 1 - (entry['odds'] - min_odds) / (max_odds - min_odds)

        win_odds_base = round(1.5 + (1 - strength_norm) * 7.5, 2)
        draw_odds_base = round(3.0 + abs(strength_norm - 0.5) * 0.8 + random.uniform(-0.1, 0.1), 2)
        loss_odds_base = round(1.2 + strength_norm * 6.0, 2)

        entry['win_odds_base'] = win_odds_base
        entry['draw_odds_base'] = draw_odds_base
        entry['loss_odds_base'] = loss_odds_base

    # Write back to file
    with open(odds_file, 'w', encoding='utf-8') as f:
        json.dump(odds_data, f, indent=2, ensure_ascii=False)

    print(f"Extended {len(odds_data)} entries in odds.json")

def main():
    """Main function to generate all pre-match data"""
    print("Generating pre-match data...")

    # Get team names
    team_codes = get_team_names()
    print(f"Found {len(team_codes)} teams: {team_codes}")

    # Generate strategy.json and recent_form.json for each team
    for team_id in team_codes:
        # Get team name directory
        team_name_dir = None
        for item in os.listdir("my-world-cup/database/2_ability_models"):
            if os.path.isdir(f"my-world-cup/database/2_ability_models/{item}"):
                # Check if this directory corresponds to our team
                if ' ' in item:
                    code = item.split()[-1]
                else:
                    code = item[:3].upper()

                if code == team_id:
                    team_name_dir = item
                    break

        if not team_name_dir:
            print(f"Warning: Could not find directory for team {team_id}")
            continue

        # Get overall score
        overall_score = get_team_overall_score(team_name_dir)

        # Generate strategy
        strategy = generate_strategy(team_id, overall_score)
        strategy_file = f"my-world-cup/database/2_ability_models/{team_name_dir}/strategy.json"

        with open(strategy_file, 'w', encoding='utf-8') as f:
            json.dump(strategy, f, indent=2, ensure_ascii=False)

        # Generate recent form
        recent_form = generate_recent_form(team_id, overall_score)
        form_file = f"my-world-cup/database/2_ability_models/{team_name_dir}/recent_form.json"

        with open(form_file, 'w', encoding='utf-8') as f:
            json.dump(recent_form, f, indent=2, ensure_ascii=False)

        print(f"Generated data for {team_id} ({team_name_dir}): overall={overall_score}")

    # Extend odds file
    extend_odds_file()

    print("Pre-match data generation complete!")

if __name__ == "__main__":
    main()